File: fluentcore/src/main/java/com/mcreater/fxfluent/controls/FluentProgressBar.java
package com.mcreater.fxfluent.controls;

import com.mcreater.fxfluent.controls.abstractions.Backgroundable;
import com.mcreater.fxfluent.controls.abstractions.Dictable;
import com.mcreater.fxfluent.controls.abstractions.Foregroundable;
import com.mcreater.fxfluent.controls.abstractions.Themeable;
import com.mcreater.fxfluent.controls.skin.FluentProgressBarSkin;
import com.mcreater.fxfluent.controls.value.StateMap;
import com.mcreater.fxfluent.xaml.ResourceDict;
import com.mcreater.fxfluent.xaml.XamlManager;
import com.mcreater.fxfluent.xaml.style.AppColorTheme;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.scene.control.ProgressBar;
import javafx.scene.control.Skin;

import static com.mcreater.fxfluent.controls.value.ControlMaps.HyperLinkButton.Companion;

public class FluentProgressBar extends ProgressBar implements Backgroundable, Foregroundable, Dictable, Themeable {
    private AppColorTheme theme = AppColorTheme.SYSTEM;
    public enum IndeterminateState {
        NORMAL, PAUSE, ERROR
    }
    private final ObjectProperty<IndeterminateState> indeterminateState = new SimpleObjectProperty<>(IndeterminateState.NORMAL);
    public ObjectProperty<IndeterminateState> indeterminateState() {
        return indeterminateState;
    }

    public IndeterminateState getIndeterminateState() {
        return indeterminateState().get();
    }
    public void setIndeterminateState(IndeterminateState state) {
        indeterminateState().set(state);
    }

    public FluentProgressBar() {
        init();
    }

    public FluentProgressBar(double progress) {
        super(progress);
        init();
    }

    private void init() {
        getStyleClass().add("fluent-progress-bar");
    }

    public String getUserAgentStylesheet() {
        return FluentProgressBar.class.getClassLoader().getResource("css/FluentProgressBar.css").toString();
    }

    protected Skin<?> createDefaultSkin() {
        return new FluentProgressBarSkin(this);
    }

    public StateMap getBackgroundRemap() {
        return Companion.getBG_KEY_MAP();
    }

    public StateMap getForegroundRemap() {
        return Companion.getFG_KEY_MAP();
    }

    public ResourceDict getResourceDict() {
        return XamlManager.getDict(theme);
    }

    public void onUpdateTheme(AppColorTheme theme) {
        this.theme = theme;
        ((FluentProgressBarSkin) getSkin()).implUpdate();
    }
}

File: fluentcore/src/main/java/com/mcreater/fxfluent/controls/FluentProgressBar.kt
package com.mcreater.fxfluent.controls

import com.mcreater.fxfluent.controls.abstractions.*
import com.mcreater.fxfluent.controls.skin.FluentProgressBarSkin
import com.mcreater.fxfluent.controls.value.ControlMaps.ProgressBar.Companion.BG_KEY_MAP
import com.mcreater.fxfluent.controls.value.ControlMaps.ProgressBar.Companion.FG_KEY_MAP
import com.mcreater.fxfluent.controls.value.StateMap
import com.mcreater.fxfluent.xaml.ResourceDict
import com.mcreater.fxfluent.xaml.XamlManager.Companion.getDict
import com.mcreater.fxfluent.xaml.style.AppColorTheme
import javafx.beans.property.ObjectProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.scene.control.ProgressBar
import javafx.scene.control.Skin


open class FluentProgressBar : ProgressBar, Backgroundable, Foregroundable, Dictable, Themeable {
    private var theme: AppColorTheme? = AppColorTheme.SYSTEM

    enum class IndeterminateState {
        NORMAL,
        PAUSE,
        ERROR
    }

    private val indeterminateStateProperty: ObjectProperty<IndeterminateState> = SimpleObjectProperty(IndeterminateState.NORMAL)
    var indeterminateState: IndeterminateState
        get() = indeterminateStateProperty.get()
        set(value) = indeterminateStateProperty.set(value)
    fun indeterminateStateProperty(): ObjectProperty<IndeterminateState> {
        return indeterminateStateProperty
    }

    constructor() {
        init()
    }

    constructor(progress: Double) : super(progress) {
        init()
    }

    private fun init() {
        styleClass.add("fluent-progress-bar")
    }

    override fun getUserAgentStylesheet(): String {
        return FluentProgressBar::class.java.getClassLoader().getResource("css/FluentProgressBar.css")?.toString() ?: ""
    }

    override fun createDefaultSkin(): Skin<*> {
        return FluentProgressBarSkin(this)
    }

    override val backgroundRemap: StateMap
        get() = BG_KEY_MAP
    override val foregroundRemap: StateMap
        get() = FG_KEY_MAP
    override val resourceDict: ResourceDict
        get() = getDict(theme!!)

    override fun onUpdateTheme(theme: AppColorTheme?) {
        this.theme = theme
        (skin as SkinUpdatable).implUpdate()
    }
}
